Reimplement HTTP service check program in Python for better redirect handling

sysvinit opened the pull request:
The `check_http` program in monitoring-plugins which is used for executing customer-provided HTTP service checks does not provide a way to check for specific status codes after following redirects. The only available behaviour is: (a) follow redirects, but then evaluate the resulting status code using the default built-in behaviour; (b) check the status code against the specified list of expected status codes, but do not follow redirects.

This change provides a minimal Python reimplementation of the HTTP check program in fc-sensuplugins. This script supports checking the status codes against an explicitly specified list both with and without following any redirects first.

The `fc-monitor` script has correspondingly been updated to generate sensu checks which use the Python script instead of the monitoring-plugins program.

PL-132208

@flyingcircusio/release-managers

## Release process

Impact: internal

Changelog:

### PR release workflow (internal)

- [x] PR has internal ticket
- [x] internal issue ID (PL-…) part of branch name
- [x] internal issue ID mentioned in PR description text
- [x] ticket is on Platform agile board
- [x] ticket state set to *Pull request ready*
- [ ] if ticket is more urgent than within the next few days, directly contact a member of the Platform team

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - The redirect handling for customer-provided HTTP checks should observe the principle of least astonishment when combined with an explicit list of expected status codes.
- [x] Security requirements tested? (EVIDENCE)
  - Tested under global maintenance on services15. The script runs properly and check results are populated correctly in Sensu.

sysvinit committed:
sensuplugins: add python reimplementation of check_http

PL-132208

sysvinit committed:
agent: use python http check script for customer service checks

PL-132208

ctheune commented on pkgs/fc/sensuplugins/fc/sensuplugins/http.py:
```diff
@@ -0,0 +1,131 @@
+    try:
+        resp = session.get(target, allow_redirects=args.follow)
+    except requests.ConnectionError as ex:
+        critical("could not connect to remote host")
```
adding exception detail here would be great, or at least the specific connection error message

sysvinit replied:
Fixed in [39ac04b](https://github.com/flyingcircusio/fc-nixos/pull/940/commits/39ac04b0256307c4a9dc61bd0c4b9d08e9284798).

ctheune commented on pkgs/fc/sensuplugins/fc/sensuplugins/http.py:
```diff
@@ -0,0 +1,131 @@
+    except requests.ConnectionError as ex:
+        critical("could not connect to remote host")
+    except requests.TooManyRedirects as ex:
+        warning("maximum redirection depth exceeded")
```
I think this could/should be critical.

sysvinit replied:
This intentionally matches the behaviour of monitoring-plugins (https://github.com/monitoring-plugins/monitoring-plugins/blob/v2.3.5/plugins/check_http.c#L1599-L1602). Should we keep this behaviour?

ctheune replied:
nope. let's make this critical.

ctheune replied:
So, browsers have two behaviours that we can detect and predict whether this page will end up as an error in a user's browser:

1. They follow a limited number of redirects, before displaying an error page. Chrome and Firefox seem to be at 20 by default. IE8 supposedly was limiting it to 10.
2. They seem to detect redirection loops, so that could be an improvement for the future.

I'd be happy if we stay with 10 as a limit for now and make a note about this as well as the potential improvement to detect loops.

ctheune replied:
So, browsers have two behaviours that we can detect and predict whether this page will end up as an error in a user's browser:

1. They follow a limited number of redirects, before displaying an error page. Chrome and Firefox seem to be at 20 by default. IE8 supposedly was limiting it to 10.
2. They seem to detect redirection loops, so that could be an improvement for the future.

I'd be happy if we stay with 10 as a limit for now and make a note about this as well as the potential improvement to detect loops.

sysvinit replied:
See [88b4e75](https://github.com/flyingcircusio/fc-nixos/pull/940/commits/88b4e75496fa600d2d6641655a06b096faf1932f) and [48b0f76](https://github.com/flyingcircusio/fc-nixos/pull/940/commits/48b0f764d57285d3463383612333e3a647ce217a).

ctheune commented on pkgs/fc/sensuplugins/fc/sensuplugins/http.py:
```diff
@@ -0,0 +1,131 @@
+    except requests.TooManyRedirects as ex:
+        warning("maximum redirection depth exceeded")
+    except requests.Timeout as ex:
+        critical("request to remote host timed out")
```
adding exception detail here would be great, or if we can at least state the specific timeout threshold that was crossed

sysvinit replied:
`requests` doesn't actually set any timeouts by default, which I missed when I skimmed through the documentation earlier this week, so currently this doesn't actually do anything. But yes, I'll add timeout configuration and put that into the message.

sysvinit replied:
Fixed in [29b7899](https://github.com/flyingcircusio/fc-nixos/pull/940/commits/29b78996ebd2da0f16565d6363d9eacb29d910fe).

ctheune commented on pkgs/fc/sensuplugins/fc/sensuplugins/http.py:
```diff
@@ -0,0 +1,131 @@
+                f"request returned server error status code: {resp.status_code}"
+            )
+        elif resp.status_code >= 400:
+            warning(
```
this should be critical, too

sysvinit replied:
This is likewise behaviour adopted from monitoring-plugins (https://github.com/monitoring-plugins/monitoring-plugins/blob/v2.3.5/plugins/check_http.c#L1217-L1221). Do we want to keep this behaviour too?

ctheune replied:
In general a 400 error code indicates an error on the client side. However: in monitoring we create specific requests that are supposed NOT to be in error and thus a change on the server side (e.g. a URL going away) can cause an error that otherwise would correctly be an error on the client side. In this case it becomes: either your monitoring is out of date, or we accidentally broke something on the server side. We can't distinguish this and should err on the side of caution and consider this critical.

sysvinit replied:
Fixed in [48b0f76](https://github.com/flyingcircusio/fc-nixos/pull/940/commits/48b0f764d57285d3463383612333e3a647ce217a).

sysvinit committed:
sensuplugins: configure timeouts in http check script

PL-132208

sysvinit committed:
sensuplugins: add error details to message in http script

PL-132208

sysvinit committed:
sensuplugins: update http exit codes to match sensu plugin spec

PL-132208

sysvinit committed:
sensuplugins: expand commentary on redirection limit behaviour

PL-132208

sysvinit committed:
sensuplugins: change warnings to critical errors

This better matches the desired semantics from a service monitoring
perspective.

PL-132208

sysvinit committed:
servicecheck: wrap http check script separately

sensu-client needs check_http_service in its PATH for the directory
service checks, however adding fc-sensuplugins in its entirety causes
collisions with binaries provided by monitoring-plugins. Import only
the check script with another derivation for indirection.

PL-132208
